lib/toolbox/stream/string_stream.c

```diff
@@ -0,0 +1,182 @@
+static size_t string_stream_tell(StringStream* stream);
+static size_t string_stream_size(StringStream* stream);
+static size_t string_stream_write(StringStream* stream, const char* data, size_t size);
+static size_t string_stream_read(StringStream* stream, char* data, size_t size);
+static bool string_stream_delete_and_insert(
+    StringStream* stream,
+    size_t delete_size,
+    StreamWriteCB write_callback,
+    const void* ctx);
+
+const StreamVTable string_stream_vtable = {
+    .free = (StreamFreeFn)string_stream_free,
+    .eof = (StreamEOFFn)string_stream_eof,
+    .clean = (StreamCleanFn)string_stream_clean,
+    .seek = (StreamSeekFn)string_stream_seek,
+    .tell = (StreamTellFn)string_stream_tell,
+    .size = (StreamSizeFn)string_stream_size,
+    .write = (StreamWriteFn)string_stream_write,
+    .read = (StreamReadFn)string_stream_read,
+    .delete_and_insert = (StreamDeleteAndInsertFn)string_stream_delete_and_insert,
+};
+
+Stream* string_stream_alloc() {
+    StringStream* stream = malloc(sizeof(StringStream));
+    string_init(stream->string);
+    stream->index = 0;
+    stream->stream_base.vtable = &string_stream_vtable;
+    return (Stream*)stream;
+}
+
+static void string_stream_free(StringStream* stream) {
+    string_clear(stream->string);
+    free(stream);
+}
+
+static bool string_stream_eof(StringStream* stream) {
+    return (string_stream_tell(stream) >= string_stream_size(stream));
+}
+
+static void string_stream_clean(StringStream* stream) {
+    stream->index = 0;
+    string_reset(stream->string);
+}
+
+static bool string_stream_seek(StringStream* stream, int32_t offset, StreamOffset offset_type) {
+    bool result = true;
+    switch(offset_type) {
+    case StreamOffsetFromStart:
+        if(offset >= 0) {
+            stream->index = offset;
+        } else {
+            result = false;
+            stream->index = 0;
+        }
+        break;
+    case StreamOffsetFromCurrent:
+        if(((int32_t)stream->index + offset) >= 0) {
+            stream->index += offset;
+        } else {
+            result = false;
+            stream->index = 0;
+        }
+        break;
+    case StreamOffsetFromEnd:
+        if(((int32_t)string_size(stream->string) + offset) >= 0) {
+            stream->index = string_size(stream->string) + offset;
+        } else {
+            result = false;
+            stream->index = 0;
+        }
+        break;
+    }
+
+    int32_t diff = (stream->index - string_size(stream->string));
+    if(diff > 0) {
+        stream->index -= diff;
+        result = false;
+    }
+
+    return result;
+}
+
+static size_t string_stream_tell(StringStream* stream) {
+    return stream->index;
+}
+
+static size_t string_stream_size(StringStream* stream) {
+    return string_size(stream->string);
+}
+
+static size_t string_stream_write(StringStream* stream, const char* data, size_t size) {
+    // TODO: can be optimized for edge cases
```

Comment:
at least do reserve space for already-known size

Reply:
I left it for future optimizations